Add citybikes platform

aronsky opened the pull request:
## Description:

[CityBikes](https://citybik.es/#about) is an open API platform, that provides data about bike sharing systems around the world. It allows real-time monitoring of bike availability at bike sharing stations.

This platform connects the CityBikes API to home assistant, opening up the possibilities for advanced automations, such as recommending the user a bike sharing station based on the amount of bikes available, upon leaving the house.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#2770

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: citybikes
    radius: 500
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

aronsky committed:
Initial commit - new CityBikes platform

aronsky committed:
Several syntax fixes.

aronsky committed:
Added imperial unit support.

aronsky committed:
Added station list lenght validation.

aronsky committed:
Style fixes.

aronsky committed:
Updated requirements.

aronsky committed:
Updated .coveragerc.

aronsky committed:
Fixed style problems according to pylint output.

balloob commented on homeassistant/components/sensor/citybikes.py:
```diff
@@ -0,0 +1,186 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = timedelta(seconds=12)  # Arbitrary
```
Let's make this 5 minutes, 12 seconds is way too frequent. We'll hammer their API.

aronsky replied:
Agreed, I will fix it when I'm back at my laptop.

balloob commented on homeassistant/components/sensor/citybikes.py:
```diff
@@ -0,0 +1,186 @@
+        self._free_bikes = station[ATTR_FREE_BIKES]
+        self._timestamp = station[ATTR_TIMESTAMP]
+        self._friendly_name = station[ATTR_STATION_NAME]
+        self._name = slugify("{}_{}".format(self._base_name,
```
Names don't need to be slugified, instead they should be a real name including spaces.

aronsky replied:
No problem, I will fix it when I'm back at my laptop.

balloob commented on homeassistant/components/sensor/citybikes.py:
```diff
@@ -0,0 +1,186 @@
+            if station[ATTR_STATION_ID] == station_id:
+                return station
+
+    @Throttle(SCAN_INTERVAL)
```
No need to throttle, by defining scan interval, the sensor component will only call update every SCAN_INTERVAL time.

aronsky replied:
What's throttled here is the `update` method of the poller object, not the station entities. Every time a `SCAN_INTERVAL` passes, each monitored station will indirectly call upon the `update` method of the caller, and the `Throttle` decorator makes sure only one actual request to the API is made.

It would be cleaner if each station could request data about itself, but unfortunately, that's not currently supported by the API.

Another option would be to create a separate entity for the whole bike sharing network, which will be updated at `SCAN_INTERVAL`, and then manually update each station entity. I'm planning to do it that way later, so that the platform is set up correctly when HASS starts offline, and the stations are added later, once connectivity is restored. But for now, I'm not sure how to achieve that correctly, hence the current approach.

rytilahti commented on homeassistant/components/sensor/citybikes.py:
```diff
@@ -0,0 +1,186 @@
+                                                longitude, stations_list))
+
+
+class CityBikesNetworkPoller(object):
```
explicit inheritance from object is not required in Python 3.

aronsky replied:
I'm used to inheriting from `object` :)
I also looked at other components, and saw it being done throughout the code, so thought it's acceptable.

balloob replied:
Yeah either is fine with me 👍

rytilahti commented:
Nice component (and service), thanks! I tested out the functioning of radius and static station lists, working nicely.

aronsky committed:
Updated SCAN_INTERVAL value.

aronsky committed:
Fixed station names.
Removed unnecessary calls to `slugify`.
Changed the base name to reflect the name of the bike sharing
network, instead of the more generic `citybikes`.

aronsky committed:
Small style fix.

aronsky commented:
@balloob,

I already have some ideas for improvements that I want to implement ASAP, and they will be breaking the current structure, so I guess there's no point in merging this just yet.

In general, I'm planning the following - let me know what you think:
1. Promote it into a component. Is that OK, or should I leave it as a sensor? I believe its role is different enough to justify such a promotion (like weather, for example).
2. Switch to direct access to CityBikes API. This will let me handle the HTTP requests, and I will be able to make the code properly async. Their API is officially open and announced to the web, so I don't think that constitutes accessing data directly instead of using standalone Python libraries. I will, however, contribute the async code to the Python library that wraps access to the data (the one I'm using currently), and once that is merged, I'll switch back to using the library.
3. Additionally, I want to support dynamic creation/removal of Station entities. Both to allow the component to start without errors while the Internet is down, and to support dynamic location - so that, for example, a user will always be able to see free bikes on stations around their current location (for users who use device location tracking). Are there any rules or guidelines about how to properly implement such a thing?

aronsky commented:
So, I wrote a component version (https://github.com/home-assistant/home-assistant/compare/dev...aronsky:add-citybikes-component). It's better in some ways, but not perfect - I'd like some help to develop it further. Even if the final decision will be to leave this as a platform, it was a good exercise in async development, and I'll port the appropriate changes to this pull request.

Anyway, what's the best way to discuss the component? Should I make another pull request, and once we decide which approach is better, you merge one and close the other? I don't want to spam the project, I'm sure you guys have enough on your heads.

balloob commented:
Let's keep the discussion here. So I prefer the async version because it's async. However, I do think that:

 - the async code should be extracted into a lib on pypi.
 - I think that this should stay a sensor as you know, that's kinda what it is.

rytilahti commented:
Is the upstream of the already existing python api responsive? I think it would make sense to try to integrate the async version to the already existing library (if doable with a feasible amount of work).

aronsky commented:
I already did, today (https://github.com/eskerda/python-citybikes/pull/1). You're welcome to review and comment there :) However, this breaks the compatibility with Python 2.7, so not sure it'll be accepted. For now, I'll use my commit as the requirement, and update it if the pull request gets accepted.

aronsky committed:
Use async version of python-citybikes

aronsky committed:
Made platform setup async.

aronsky committed:
Made some more things async.

aronsky committed:
Switched to constants.

aronsky commented:
Don't mind the intermediate pushes to the branch, it's a WIP (I'm closing this PR for now, will re-open it once I'm done).

aronsky commented:
Can't re-open this PR due to GitHub limitations (I rebased it on an up-to-date `dev` commit, and GitHub doesn't see the continuity). Continued in #8202.
